### PR TITLE
Issue #237: Improve cache management

### LINKED ISSUE #237: Side-effecting operations on `GitContext` should automatically invalidate relevant caches
For instance, `fetch_remotes` will likely make `__remote_branches_cached` and `__commit_sha_by_revision_cached` invalid. As for now, we're just selectively doing `flush_caches`... but that's not enough since there're more methods that modify the state of the underlying git repo, while not invalidating any caches.

I'm afraid some bugs might be lurking even in the command-line client, and there would surely be bugs from the PoV of programatic access.

### PR DESCRIPTION
**Pull request by BartoszSiwinski**
It's mostly method names correction in places where I found it confusing like remotes, was actually a method and not a property/instance variable.

In a `TODO` comment I placed pseudo justification for a `flush_caches()` in that place. In some cases is just an example of a cache that would be no longer valid.

### CODE REVIEW

File: git_machete/git_operations.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def get_default_editor(self) -> Optional[str]:
         proposed_editor_funs: List[Tuple[str, Callable[[], Optional[str]]]] = [
             ("$" + git_machete_editor_var, lambda: os.environ.get(git_machete_editor_var)),
```
Comment (PawelLipski) [inline git_machete/git_operations.py:65]: https://github.com/VirtusLab/git-machete/issues/251 ... something to consider while you're at this PR. Not to be done right now, but just to confirm that the existing `run_git` and `popen_git` calls from client.py are safe from the PoV of cache validity.

File: git_machete/git_operations.py [hunk 2]
```diff
@@ -167,9 +167,13 @@ def fetch_remote(self, remote: str) -> None:
         if remote not in self.__fetch_done_for:
             self.run_git("fetch", remote)
             self.__fetch_done_for.add(remote)
+            # TODO (#237): flush, remote_branches_cached and others
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: For clarity... what is supposed to happen for these TODOs before the PR is merged? 🤔

Comment (BartoszSiwinski) [reply git_machete/git_operations.py]: They're to be deleted - whole. I don't think it makes sense to keep them as a comment or smth like that. I created them for myself as a debug duck and for reviewers to help them with the understanding of the decisiones. After the review is completed/initially approved they no longer have any purpose.

File: git_machete/git_operations.py (hunk 1 above)
Comment (BartoszSiwinski) [reply git_machete/git_operations.py:65]: I looked into it and they're not. For example `delete_unmanaged` in my opinion should be flushing the cache. Cache clearing should be done on the side of the actual executor of the command that invalidates the cache. `delete_unmanaged` is defined in `client.py` and does that amongst other lines: 
```
if ans in ('y', 'yes'):
    self.__git.run_git("branch", "-d" if is_merged_to_remote else "-D", branch)
```
Ideally we would have here a call to a wrapper around `git branch -d/-D,` and this wrapper would call `run_git` and also invalidate/clear the cache. I would say that the issue #251 should be pulled into `v3.4.0 `release and as a followup/part of this story cache management should be reconsidered.

File: git_machete/git_operations.py [hunk 3]
```diff
@@ -573,14 +586,14 @@ def merged_local_branches(self) -> List[str]:
         ))
 
     def get_hook_path(self, hook_name: str) -> str:
```
Comment (PawelLipski) [inline git_machete/git_operations.py:585]: While you're here... `merged_local_branches` -> `get_merged_local_branches`

File: git_machete/git_operations.py [hunk 4]
```diff
@@ -405,15 +413,20 @@ def create_branch(self, branch: str, out_of_revision: str) -> None:
         self.flush_caches()  # the repository state has changed because of a successful branch creation, let's defensively flush all the caches
```
Comment (PawelLipski) [inline git_machete/git_operations.py:410]: One nuance... `checkout` AFAICS can technically invalidate `__reflogs_cached` and all the `__..._by_revision_cached`... but under the single key,  `HEAD` 🤔

Comment (PawelLipski) [reply git_machete/git_operations.py:410]: The question is... are any of this caches used for the `HEAD` key?

Comment (PawelLipski) [reply git_machete/git_operations.py:410]: Even if they aren't currently... there's (probably?) currently nothing in place that would forbid such usage

File: git_machete/git_operations.py (hunk 1 above)
Comment (PawelLipski) [reply git_machete/git_operations.py:65]: Okay! into v3.4.0 it goes then

File: git_machete/git_operations.py (hunk 4 above)
Comment (BartoszSiwinski) [reply git_machete/git_operations.py:410]: > One nuance... `checkout` AFAICS can technically invalidate `__reflogs_cached` and all the `__..._by_revision_cached`... but under the single key, `HEAD` 🤔

How can it invalidate that cache?

Comment (PawelLipski) [reply git_machete/git_operations.py:410]: Just `flush_caches()` as well, let's not over-optimize with just deleting a single key

Comment (BartoszSiwinski) [reply git_machete/git_operations.py:410]: Ok, I've added a `self.flush_caches` at the end of `GitContext.checkout` method

**Review by PawelLipski — APPROVED** [review]

**Pull request merged**